ea.com: top bar separates when scrolling down any page, causing a tear in the UI

graouts opened the pull request:
#### 9de4589d1aab3070929e8425c4a12babb5dd6c8b
<pre>
ea.com: top bar separates when scrolling down any page, causing a tear in the UI
<a href="https://bugs.webkit.org/show_bug.cgi?id=294840">https://bugs.webkit.org/show_bug.cgi?id=294840</a>
<a href="https://rdar.apple.com/106770785">rdar://106770785</a>

Reviewed by Anne van Kesteren.

There are two issues here:

1. paint performance is suboptimal as we scroll and we take upwards of 18ms,
2. we are running separate animations that we would hope to be in sync but cannot be because
   one is accelerated and the other isn’t and a gap shows. Due to our architecture for running
   such animations, one will run at 120fps and the other one will top up at 60fps.

We address this bug by introducing a quirk to disable the `transform` accelerated animation
targeting the `&lt;ea-network-nav&gt;` element on all ea.com pages.

* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::canBeAccelerated const):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldPreventKeyframeEffectAcceleration const):
(WebCore::handleEAQuirks):
(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:

Canonical link: <a href="https://commits.webkit.org/296517@main">https://commits.webkit.org/296517@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/46a1686251f0b3e558b870ac189986c94ef5cbf2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108772 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28433 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18857 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113982 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59151 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110735 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29122 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36997 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82644 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111720 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23140 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97976 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63082 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22562 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16114 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58683 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/92510 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16162 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117104 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35826 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/26457 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91663 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36199 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94243 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91470 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36370 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14124 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31710 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17561 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35726 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41260 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35436 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38782 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37114 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1894,6 +1895,19 @@ bool Quirks::shouldRewriteMediaRangeRequestForURL(const URL& url) const
+    if (!target)
+        return false;
+
+    return Ref { target->element }->localName() == "ea-network-nav"_s;
```
```suggestion
    return target && Ref { target->element }->localName() == "ea-network-nav"_s;
```

annevk commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -2386,6 +2400,13 @@ static void handleESPNQuirks(QuirksData& quirksData, const URL& quirksURL, const
+{
+    UNUSED_PARAM(quirksURL);
+    UNUSED_PARAM(documentURL);
+    quirksData.isEA = quirksDomainString == "ea.com"_s;
```
I'm pretty sure this is not what we want to do because now we assign `isEA` all the time. You want to follow the early return pattern that the other methods have.

graouts committed:
ea.com: top bar separates when scrolling down any page, causing a tear in the UI
https://bugs.webkit.org/show_bug.cgi?id=294840
rdar://106770785

Reviewed by Anne van Kesteren.

There are two issues here:

1. paint performance is suboptimal as we scroll and we take upwards of 18ms,
2. we are running separate animations that we would hope to be in sync but cannot be because
   one is accelerated and the other isn’t and a gap shows. Due to our architecture for running
   such animations, one will run at 120fps and the other one will top up at 60fps.

We address this bug by introducing a quirk to disable the `transform` accelerated animation
targeting the `<ea-network-nav>` element on all ea.com pages.

* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::KeyframeEffect::canBeAccelerated const):
* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldPreventKeyframeEffectAcceleration const):
(WebCore::handleEAQuirks):
(WebCore::Quirks::determineRelevantQuirks):
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:

Canonical link: https://commits.webkit.org/296517@main
